### ISSUE TITLE
NumPy API surface : plan/prioritize the coverage

### ISSUE THREAD

**Issue by ev-br** [Quansight-Labs/numpy_pytorch_interop#87]
Splitting it off gh-86, here's the difference in API surfaces of NumPy and this wrapper. We can edit the order  to reflect priorities:

```
>>> import numpy as np
>>> import torch_np as tnp
>>> npset = set(x for x in dir(np) if not x.startswith('_') and not inspect.ismodule(x) and not x[0].isupper())
>>> tnpset = set(dir(tnp))
>>> for name in sorted(npset - tnpset):
...        print("-[ ]", name)
```

-[ ] add_docstring
-[ ] add_newdoc
-[ ] add_newdoc_ufunc
-[ ] append
-[ ] apply_along_axis
-[ ] apply_over_axes
-[ ] argpartition
-[ ] array2string
-[ ] array_repr
-[ ] array_str
-[ ] asanyarray
-[ ] asarray_chkfinite
-[ ] ascontiguousarray
-[ ] asfortranarray
-[ ] asmatrix
-[ ] bartlett
-[ ] base_repr
-[ ] binary_repr
-[ ] bitwise_not
-[ ] blackman
-[ ] block
-[ ] bmat
-[ ] broadcast
-[ ] busday_count
-[ ] busday_offset
-[ ] busdaycalendar
-[ ] byte_bounds
-[ ] bytes_
-[ ] c_
-[ ] cast
-[ ] cfloat
-[ ] char
-[ ] character
-[ ] chararray
-[ ] choose
-[ ] clongdouble
-[ ] clongfloat
-[ ] compare_chararrays
-[ ] compat
-[ ] complex256
-[ ] complex_
-[ ] compress
-[ ] convolve
-[ ] copyto
-[ ] correlate
-[ ] cross
-[ ] ctypeslib
-[ ] datetime64
-[ ] datetime_as_string
-[ ] datetime_data
-[ ] delete
-[ ] deprecate
-[ ] deprecate_with_doc
-[ ] digitize
-[ ] disp
-[ ] e
-[ ] ediff1d
-[ ] einsum
-[ ] einsum_path
-[ ] emath
-[ ] euler_gamma
-[ ] extract
-[ ] fastCopyAndTranspose
-[ ] fft
-[ ] find_common_type
-[ ] fix
-[ ] flatiter
-[ ] flexible
-[ ] float128
-[ ] format_float_positional
-[ ] format_float_scientific
-[ ] format_parser
-[ ] frexp
-[ ] from_dlpack
-[ ] frombuffer
-[ ] fromfile
-[ ] fromfunction
-[ ] fromiter
-[ ] frompyfunc
-[ ] fromregex
-[ ] fromstring
-[ ] genfromtxt
-[ ] get_array_wrap
-[ ] get_include
-[ ] get_printoptions
-[ ] getbufsize
-[ ] geterr
-[ ] geterrcall
-[ ] geterrobj
-[ ] gradient
-[ ] hamming
-[ ] hanning
-[ ] histogram
-[ ] histogram2d
-[ ] histogram_bin_edges
-[ ] histogramdd
-[ ] in1d
-[ ] info
-[ ] infty
-[ ] insert
-[ ] interp
-[ ] intersect1d
-[ ] is_busday
-[ ] isfortran
-[ ] isin
-[ ] isnat
-[ ] issctype
-[ ] issubclass_
-[ ] issubsctype
-[ ] iterable
-[ ] ix_
-[ ] kaiser
-[ ] kernel_version
-[ ] lexsort
-[ ] lib
-[ ] linalg
-[ ] little_endian
-[ ] load
-[ ] loadtxt
-[ ] longcomplex
-[ ] longdouble
-[ ] longfloat
-[ ] lookfor
-[ ] ma
-[ ] mask_indices
-[ ] mat
-[ ] math
-[ ] matrix
-[ ] maximum_sctype
-[ ] may_share_memory
-[ ] memmap
-[ ] mgrid
-[ ] min_scalar_type
-[ ] mod
-[ ] msort
-[ ] nbytes
-[ ] ndenumerate
-[ ] ndindex
-[ ] nditer
-[ ] nested_iters
-[ ] numarray
-[ ] obj2sctype
-[ ] object_
-[ ] ogrid
-[ ] oldnumeric
-[ ] packbits
-[ ] pad
-[ ] partition
-[ ] piecewise
-[ ] place
-[ ] poly
-[ ] poly1d
-[ ] polyadd
-[ ] polyder
-[ ] polydiv
-[ ] polyfit
-[ ] polyint
-[ ] polymul
-[ ] polynomial
-[ ] polysub
-[ ] polyval
-[ ] printoptions
-[ ] promote_types
-[ ] put
-[ ] putmask
-[ ] r_
-[ ] rec
-[ ] recarray
-[ ] recfromcsv
-[ ] recfromtxt
-[ ] record
-[ ] require
-[ ] resize
-[ ] roots
-[ ] safe_eval
-[ ] save
-[ ] savetxt
-[ ] savez
-[ ] savez_compressed
-[ ] sctype2char
-[ ] sctypeDict
-[ ] select
-[ ] set_numeric_ops
-[ ] set_printoptions
-[ ] set_string_function
-[ ] setbufsize
-[ ] setdiff1d
-[ ] seterr
-[ ] seterrcall
-[ ] seterrobj
-[ ] setxor1d
-[ ] shares_memory
-[ ] show_config
-[ ] show_runtime
-[ ] singlecomplex
-[ ] sort_complex
-[ ] source
-[ ] spacing
-[ ] str_
-[ ] string_
-[ ] take
-[ ] tensordot
-[ ] test
-[ ] testing
-[ ] timedelta64
-[ ] tracemalloc_domain
-[ ] trapz
-[ ] trim_zeros
-[ ] true_divide
-[ ] typename
-[ ] ufunc
-[ ] uint
-[ ] uint16
-[ ] uint32
-[ ] uint64
-[ ] uintc
-[ ] uintp
-[ ] ulonglong
-[ ] unicode_
-[ ] union1d
-[ ] unpackbits
-[ ] unwrap
-[ ] use_hugepage
-[ ] ushort
-[ ] vectorize
-[ ] version
-[ ] void
-[ ] who

**Comment by ev-br** [comment]
```
>>> for name in sorted(set(dir(np.ndarray)) - set(dir(tnp.ndarray))):
 ...     print("- [ ]", name)
```

- [ ] `__array__`
- [ ] `__array_finalize__`
- [ ] `__array_function__`
- [ ] `__array_interface__`
- [ ] `__array_prepare__`
- [ ] `__array_priority__`
- [ ] `__array_struct__`
- [ ] `__array_ufunc__`
- [ ] `__array_wrap__`
- [ ] `__class_getitem__`
- [ ] `__contains__`
- [ ] `__copy__`
- [ ] `__deepcopy__`
- [ ] `__delitem__`
- [ ] `__dlpack__`
- [ ] `__dlpack_device__`
- [ ] `__imatmul__`
- [ ] `__iter__`
- [ ] `__matmul__`
- [ ] `__rdivmod__`
- [ ] `__rmatmul__`
- [ ] `__setstate__`
- [ ] argpartition
- [ ] base
- [ ] byteswap
- [ ] choose
- [ ] compress
- [ ] ctypes
- [ ] data
- [ ] dump
- [ ] dumps
- [ ] fill
- [ ] flat
- [ ] getfield
- [ ] item
- [ ] itemset
- [ ] nbytes
- [ ] newbyteorder
- [ ] partition
- [ ] put
- [ ] resize
- [ ] setfield
- [ ] setflags
- [ ] take
- [ ] tobytes
- [ ] tofile
- [ ] tostring
- [ ] view

**Comment by ev-br** [comment]
For `tnp.random`:

```
>>> for name in sorted(dir(tnp.random)):
...     print("- [x]", name)
```

- [x] ArrayLike
- [x] NDArray
- [x] Optional
- [x] _default_dtype
- [x] _dtypes_impl
- [x] _helpers
- [x] _util
- [x] array_or_scalar
- [x] choice
- [x] normal
- [x] normalizer
- [x] rand
- [x] randint
- [x] randn
- [x] random
- [x] random_sample
- [x] sample
- [x] seed
- [x] shuffle
- [x] sqrt
- [x] torch
- [x] uniform



```
>>> for name in sorted(set(dir(np.random)) - set(dir(tnp.random))):
...     print("- [ ]", name)
```

- [ ] BitGenerator
- [ ] Generator
- [ ] MT19937
- [ ] PCG64
- [ ] PCG64DXSM
- [ ] Philox
- [ ] RandomState
- [ ] SFC64
- [ ] SeedSequence
- [ ] __RandomState_ctor
- [ ] `__path__`
- [ ] _bounded_integers
- [ ] _common
- [ ] _generator
- [ ] _mt19937
- [ ] _pcg64
- [ ] _philox
- [ ] _pickle
- [ ] _sfc64
- [ ] beta
- [ ] binomial
- [ ] bit_generator
- [ ] bytes
- [ ] chisquare
- [ ] default_rng
- [ ] dirichlet
- [ ] exponential
- [ ] f
- [ ] gamma
- [ ] geometric
- [ ] get_bit_generator
- [ ] get_state
- [ ] gumbel
- [ ] hypergeometric
- [ ] laplace
- [ ] logistic
- [ ] lognormal
- [ ] logseries
- [ ] mtrand
- [ ] multinomial
- [ ] multivariate_normal
- [ ] negative_binomial
- [ ] noncentral_chisquare
- [ ] noncentral_f
- [ ] pareto
- [ ] permutation
- [ ] poisson
- [ ] power
- [ ] random_integers
- [ ] ranf
- [ ] rayleigh
- [ ] set_bit_generator
- [ ] set_state
- [ ] standard_cauchy
- [ ] standard_exponential
- [ ] standard_gamma
- [ ] standard_normal
- [ ] standard_t
- [ ] test
- [ ] triangular
- [ ] vonmises
- [ ] wald
- [ ] weibull
- [ ] zipf

**Comment by ev-br** [comment]
```
>>> for name in [x for x in dir(np.linalg) if not x.startswith("_")]:
 ...     print("- [ ]", name)
``` 

- [ ] LinAlgError
- [ ] cholesky
- [ ] cond
- [ ] det
- [ ] eig
- [ ] eigh
- [ ] eigvals
- [ ] eigvalsh
- [ ] inv
- [ ] linalg
- [ ] lstsq
- [ ] matrix_power
- [ ] matrix_rank
- [ ] multi_dot
- [ ] norm
- [ ] pinv
- [ ] qr
- [ ] slogdet
- [ ] solve
- [ ] svd
- [ ] tensorinv
- [ ] tensorsolve
- [ ] test

**Comment by ev-br** [comment]
For random, we can rather easily mock up `RandomState` and `default_rng`. Might make it easier for scikit-learn and others who frown on `np.random.random` usage.

**Comment by lezcano** [comment]
I agree with the list from Ralf. Here's a slightly more comprehensive list of things that either have PyTorch equivs or are close to trivial to implement.

These could go into a PR, as most of them should have a very simple implementation. After these, I think we can declare victory on the coverage end and we should spend some time finishing the refactorisation and doing general cleanups across the codebase (without spending too much time on this) and then move on to the testing part of the project, where we show that what we built, in fact, works.

```
append (we don't have an equiv, but should be easy to implement via advance indexing)
barlett
blackman
choose (didn't we have this one already?)
copyto (funnily enough this is the function we want to use to implement the out kwarg)
convolve
cross
e
einsum
fft
from_dlpack
gradient
hamming
histogram
histogram2d (dispatch to histogram2d)
histogramdd
infty
kaiser
min_scalar_type (trivial to implement or just vendor)
nbytes (this is a property of teh tensor in PyTorch)
put
resize
take
tensordot
np.linalg
```

**Comment by rgommers** [comment]
That list seems reasonable, minus `infty` - I'm going to deprecate that one in NumPy soon, so would prefer to leave it out here.
